### PR TITLE
feat: Split FieldParsingError into different classes

### PR DESCRIPTION
**Pull request by aababilov**
Format requirements for different fields are different, so that probably
does not make sense to group them all at the same error class.

Separate error classes may be placed in different sections of a
validation report that add details on the expected format.

Users of GTFS Validator may want to reduce the severity of certain
errors, e.g., validation of URLs and phone numbers.

### CODE REVIEW

**Comment by aababilov** [comment]
@barbeau @lionel-nj 
How do you like this idea?

**Review by lionel-nj — CHANGES_REQUESTED** [review]
Great idea! That will clarify the validation reports. Just one comment inline. Thanks @aababilov!

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.mobilitydata.gtfsvalidator.notice;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A field cannot be parsed as time.
+ *
+ * <p>Time must be in the {@code HH:MM:SS} or {@code H:MM:SS} format.
```
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java:24]: We also support `HHH:MM:SS` in `GtfsTime`

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java [hunk 2]
```diff
@@ -20,12 +20,14 @@
 
 /** ID value contains something different from printable ASCII characters. */
```
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java:21]: Should have a similar comment header as other notices with more description?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java [hunk 3]
```diff
@@ -268,218 +163,181 @@ static boolean hasOnlyPrintableAscii(String s) {
...
+        columnIndex,
+        required,
+        s -> PhoneNumberUtil.getInstance().isPossibleNumber(s, feedName.getISOAlpha2CountryCode()),
+        InvalidPhoneNumberNotice::new);
   }
 
   @Nullable
   public Locale asLanguageCode(int columnIndex, boolean required) {
-    return languageCodeParser.parseField(columnIndex, required);
+    // FIXME: Enhance checks for IETF BCP 47 language code.
```
Comment (barbeau) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:199]: @aababilov Could you open an issue on GitHub for this with a better description of how this should be enhanced?

**Review by barbeau — CHANGES_REQUESTED** [review]
Thanks @aababilov! I agree that this is a good idea. Some comments in-line.

File: core/src/test/java/org/mobilitydata/gtfsvalidator/parsing/RowParserTest.java [hunk 4]
```diff
@@ -173,8 +189,15 @@ public void asPhoneNumber() {
...
+  }
 
-    assertThat(createParser("au-feed", "invalid").asPhoneNumber(0, true)).isNull();
+  @Test
+  public void asPhoneInvalid() {
+    RowParser parser = createParser("us-feed", "invalid");
+    assertThat(parser.asPhoneNumber(0, true)).isNull();
+    assertThat(parser.hasParseErrorsInRow()).isTrue();
+    assertThat(parser.getNoticeContainer().getValidationNotices())
+        .containsExactly(new InvalidEmailNotice(TEST_FILENAME, 8, "column name", "invalid"));
```
[the remark this answers is not in the document]
Comment (aababilov) [reply core/src/test/java/org/mobilitydata/gtfsvalidator/parsing/RowParserTest.java:200]: That was a bug in `Notice.equals()` that did not take `getCode()` into account. Fixed.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidColorNotice.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        ImmutableMap.of(
+            "filename",
+            filename,
+            "csvRowNumber",
+            csvRowNumber,
+            "fieldName",
+            fieldName,
+            "fieldValue",
+            fieldValue),
+        SeverityLevel.ERROR);
```
[the remark this answers is not in the document]
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidColorNotice.java:43]: This is my plan: severity should be mentioned explicitly and the overloaded constructor should be removed. That makes code more readable and explicit. The reader does not need to know that `ERROR` is the default priority.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidEmailNotice.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+package org.mobilitydata.gtfsvalidator.notice;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A field contains a malformed email address.
+ *
+ * <p>Definitions for valid emails are quite vague. We perform strict validation in the upstream
+ * using the Apache Commons EmailValidator but we allow GTFS consumers to patch the severity level
+ * of that validation.
```
[the remark this answers is not in the document]
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidEmailNotice.java:26]: Done.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java (hunk 1 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java:24]: Indeed! We should fix the documentation.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java (hunk 2 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java:21]: What description do you want me to put here?

Please note that many notices (such as EmptyFileNotice) have one-line comments.

Comment (barbeau) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java:21]: To match the other data type errors, I'd add this from the "ID" definition in GTFS:

>An ID field value is an internal ID, not intended to be shown to riders, and is a sequence of any UTF-8 characters. Using only printable ASCII characters is recommended.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java (hunk 1 above)
Comment (barbeau) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java:24]: @MobilityData/transit-specs To flag this for updating GTFS reference.

Comment (barbeau) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java:24]: @aababilov To make the validator internally consistent, I'd recommend updating these Javadocs to what the `GtfsTime` field supports (i.e., add HHH:MM:SS).

File: core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java (hunk 3 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:199]: We don't actually need them, as far as I can tell. `Locale` fully supports IETF BCP 47. I added some unit tests to verify that.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java (hunk 1 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/InvalidTimeNotice.java]: Done for InvalidTimeNotice.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java (hunk 2 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NonAsciiOrNonPrintableCharNotice.java]: Good point. Done.

**Comment by aababilov** [comment]
I am glad that you liked that idea!

**Review by barbeau — APPROVED** [review]
LGTM, thanks @aababilov!

**Comment by aababilov** [comment]
Thanks for the review!

**Pull request merged**